libraries/AP_HAL_Linux/Perf.cpp

```diff
@@ -59,126 +39,150 @@ static inline uint64_t now_nsec()
+    if (perf.type != Util::PC_ELAPSED) {
+        hal.console->printf("perf_begin() called on perf_counter_t(%s) that"
+                            " is not of PC_ELAPSED type.\n",
+                            perf.name);
         return;
     }
-    if (perf_elapsed->start == 0) {
-        hal.console->printf("perf_end() called before an perf_begin() on %s.\n",
-                            perf_elapsed->base.name);
+
+    if (perf.start == 0) {
+        hal.console->printf("perf_begin() called before begin() on perf_counter_t(%s)\n",
+                            perf.name);
         return;
     }
 
-    const uint64_t elapsed = now_nsec() - perf_elapsed->start;
-    perf_elapsed->count++;
-    perf_elapsed->total += elapsed;
+    _update_count++;
+
+    const uint64_t elapsed = now_nsec() - perf.start;
+    perf.count++;
+    perf.total += elapsed;
 
-    if (perf_elapsed->least > elapsed) {
-        perf_elapsed->least =  elapsed;
+    if (perf.least > elapsed) {
+        perf.least = elapsed;
     }
 
-    if (perf_elapsed->most < elapsed) {
-        perf_elapsed->most = elapsed;
+    if (perf.most < elapsed) {
+        perf.most = elapsed;
     }
 
     /*
      * Maintain mean and variance of interval in nanoseconds
      * Knuth/Welford recursive mean and variance of update intervals (via Wikipedia)
      * Same implementation of PX4.
      */
-    const double delta_intvl = elapsed - perf_elapsed->mean;
-    perf_elapsed->mean += (delta_intvl / perf_elapsed->count);
-    perf_elapsed->m2 += (delta_intvl * (elapsed - perf_elapsed->mean));
+    const double delta_intvl = elapsed - perf.mean;
+    perf.mean += (delta_intvl / perf.count);
+    perf.m2 += (delta_intvl * (elapsed - perf.mean));
+    perf.start = 0;
 
-    perf_elapsed->start = 0;
+#ifdef HAVE_LTTNG_UST
+    perf.lttng.end();
+#endif
 }
 
-void Util::perf_count(perf_counter_t perf)
+void Perf::count(Util::perf_counter_t pc)
 {
-    struct perf_counter_count_t *perf_counter = (struct perf_counter_count_t *)perf;
+    uintptr_t idx = (uintptr_t)pc;
 
-    if (perf_counter == NULL) {
+    if (idx >= _perf_counters.size()) {
         return;
     }
 
-    if (perf_counter->base.type != PC_COUNT) {
-        hal.console->printf("perf_count() called over a perf_counter_t(%s) "
-                            "that is not of the PC_COUNT type.\n",
-                            perf_counter->base.name);
+    Perf_Counter &perf = _perf_counters[idx];
+    if (perf.type != Util::PC_COUNT) {
+        hal.console->printf("perf_begin() called on perf_counter_t(%s) that"
+                            " is not of PC_COUNT type.\n",
+                            perf.name);
         return;
     }
 
-    perf_counter->count++;
-}
+    _update_count++;
+    perf.count++;
 
+#ifdef HAVE_LTTNG_UST
+    perf.lttng.count();
 #endif
+
+}
+
+Util::perf_counter_t Perf::add(Util::perf_counter_type type, const char *name)
+{
+    if (type != Util::PC_COUNT && type != Util::PC_ELAPSED) {
```

Comment:
Would be better this way:

if (type == Util::PC_INTERVAL)
This way the comment below makes sense.